### PR TITLE
PROV: Fix EC_KEY exporters to allow domain parameter keys

### PR DESCRIPTION
**Pull request by levitte**
The provider key export functions for EC_KEY assumed that a public key
is always present, and would fail if not.  This blocks any attempt to
export a key structure with only domain parameters.

This is similar to earlier work done in EVP_PKEY_ASN1_METHODs.

### CODE REVIEW

**Comment by levitte** [comment]
The previous similar work is #11374, now merged

File: providers/implementations/keymgmt/ec_kmgmt.c [hunk 1]
```diff
@@ -165,27 +165,28 @@ int params_to_key(EC_KEY *ec, const OSSL_PARAM params[], int include_private)
...
+    if (param_priv_key != NULL && param_pub_key == NULL)
+        goto err;
+
+    if (param_pub_key == NULL)
+        if (!OSSL_PARAM_get_octet_string(param_pub_key,
+                                         (void **)&pub_key, 0, &pub_key_len)
             || (pub_point = EC_POINT_new(ecg)) == NULL
-            || !EC_POINT_oct2point(ecg, pub_point,
-                                   pub_key, pub_key_len, ctx))
+            || !EC_POINT_oct2point(ecg, pub_point, pub_key, pub_key_len, ctx)p)
```
Comment (mattcaswell) [inline providers/implementations/keymgmt/ec_kmgmt.c]: What is this stray `p` on the end

File: providers/implementations/keymgmt/ec_kmgmt.c [hunk 2]
```diff
@@ -165,27 +165,28 @@ int params_to_key(EC_KEY *ec, const OSSL_PARAM params[], int include_private)
...
-     */
-    if (param_pub_key == NULL
-            || !OSSL_PARAM_get_octet_string(param_pub_key,
-                                            (void **)&pub_key, 0, &pub_key_len)
+
+    /* OpenSSL decree: If there's a private key, there must be a public key */
+    if (param_priv_key != NULL && param_pub_key == NULL)
+        goto err;
+
+    if (param_pub_key == NULL)
```
Comment (mattcaswell) [inline providers/implementations/keymgmt/ec_kmgmt.c]: Shouldn't this be `!=`

**Comment by mattcaswell** [comment]
The title of this PR mentions DSA - but I see no DSA changes

File: providers/implementations/keymgmt/ec_kmgmt.c (hunk 1 above)
Comment (levitte) [reply providers/implementations/keymgmt/ec_kmgmt.c]: Fixedd

File: providers/implementations/keymgmt/ec_kmgmt.c (hunk 2 above)
Comment (levitte) [reply providers/implementations/keymgmt/ec_kmgmt.c]: Fixed

**Comment by levitte** [comment]
> The title of this PR mentions DSA - but I see no DSA changes

Thanks for the reminder...  yeah it turned out that DSA was already doing it right!

**Comment by levitte** [comment]
Ping

**Comment by levitte** [comment]
Some of the work done here happened as part of #11375, but some still remained, so this PR isn't entirely lost yet.

PING

File: crypto/ec/ec_backend.c [hunk 3]
```diff
@@ -76,24 +76,25 @@ int ec_key_fromdata(EC_KEY *ec, const OSSL_PARAM params[], int include_private)
     if (ecg == NULL)
         return 0;
 
-    param_priv_key =
-        OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PRIV_KEY);
     param_pub_key =
         OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PUB_KEY);
+    if (include_private)
+        param_priv_key =
```
Comment (slontis) [inline crypto/ec/ec_backend.c]: you need to set this to NULL when it is inited..

Comment (levitte) [reply crypto/ec/ec_backend.c]: Thanks!  Fixed

**Pull request merged**